### ISSUE TITLE
Implement an ItemGroup of required properties

### ISSUE THREAD

**Issue by jeffkl** [CommonBuildToolset/CBT.Core#39]
# Overview

We want to to fail the build if one or more required properties aren't set.  Ideally, other stuff in the build graph could add on to the list as well.  This was suggested by @kingerja.

# Design
A rough design would be to have an ItemGroup that contains properties:

```xml
<ItemGroup>
    <RequiredProperty Include="MyProperty">
    <RequiredProperty Include="MyOtherProperty">
        <CustomMessage>My custom message to display if the property isn't set.</CustomMessage>
        <Code>FOO123</Code>
    </RequiredProperty>
</ItemGroup>
```

Then an `InitialTarget` would generate an error for each missing required property which would fail the build early.  Each error would need to `ContinueOnError` and a final error would actually stop build execution.

``` xml
<Project InitialTargets="CheckRequiredProperties">
    <Target Name="CheckRequiredProperties">
        <!-- Check the @(RequiredProperty) items and print error messages -->
    </Target>
</Project>
```

Design questions:
1. Should the Item name be more unique?
2. Should the target run a task for more control over the output?
3. What other metadata should be associated with the properties?  Code, custom message, etc.

Supersedes #6

**Comment by kingerja** [comment]
This feature is basically a substitute for vNext ProjectCop.  This being the case we should keep it simple and not go crazy with it.  Effectively taking the stance that anything more then checking it is defined and reporting a custom error and code should be done via projectcop.

1. Yes Generic names in a build system that is designed to be used in a totally open and uncontrolled manner will cause lots of frustration.  We may want to consider scrubbing every MSBuild property we define and prepending CBT_ to it.  Otherwise we will end up with collisions like I already do frequently in CBTv1.  
2.  Running a task introduces cross platform concerns and additional maintenance cost.  Which would be an acceptable trade off if we didn't envision having projectcop down the line where really this feature is a projectcop rule.
3.  Code, Custom Message, is as far as I would go.  I would however add an Item Definition group to act as default values if these are not specified.

**Comment by jeffkl** [comment]
Fixed in e60f12ecd7296b1f30f74da317e175965350661f
